src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -639,11 +639,69 @@ def _process_batch_upload(
     return upload_results
 
 
+def _download_assembly_fasta_file(
+        asu_client: AssemblyUtil,
+        genome_obj_info: list[Any],
+        coll_src_assembly_path: str,
+        job_data_dir: Path | str,
+) -> Path:
+    """
+    Download assembly fasta file from the workspace using AssemblyUtil.
+    Copy the assembly file from the job directory to the collection source directory.
+    """
+    try:
+        assembly_upa = genome_obj_info[10]["Assembly Object"]
+    except KeyError:
+        raise ValueError(f"Genome object {genome_obj_info[1]} does not have an assembly object linked to it")
+    containerized_assembly_file = Path(asu_client.get_assembly_as_fasta({"ref": assembly_upa})["path"])
+    # remove prefix of the assembly_path
+    relative_assembly_path = containerized_assembly_file.relative_to(_JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER)
+    local_assembly_file = Path(job_data_dir) / relative_assembly_path
+
+    loader_helper.create_hardlink_between_files(coll_src_assembly_path,
+                                                local_assembly_file)
+
+    if not os.path.exists(coll_src_assembly_path):
+        raise ValueError(f"Retrieving assembly file from Genome {genome_obj_info[1]} failed.")
+
+    return containerized_assembly_file
+
+
+def _build_assembly_tuples(
+    genome_tuples: list[WSObjTuple],
+    genome_obj_infos: list[Any],
+    asu_client: AssemblyUtil,
+    job_data_dir: str,
+) -> list[WSObjTuple]:
+    """
+    Build assembly tuples from genome tuples. Retrieve assembly fasta file from the workspace if it does not exist.
+    """
+    assembly_tuples = list()
+
+    for genome_tuple, genome_obj_info in zip(genome_tuples, genome_obj_infos):
+        assembly_file = genome_tuple.obj_name + "_assembly.fasta"  # matches the FASTA file name returned from GFU during successful upload process
+        coll_src_assembly_path = os.path.join(genome_tuple.obj_coll_src_dir, assembly_file)
+        if not os.path.exists(coll_src_assembly_path):
+            _download_assembly_fasta_file(asu_client,
+                                          genome_obj_info,
+                                          coll_src_assembly_path,
+                                          job_data_dir)
+
+        assembly_tuple = WSObjTuple(assembly_file,
+                                    genome_tuple.obj_coll_src_dir,
+                                    None)  # container_internal_file is fine to be None here as it's not used during failure recovery
```

Comment:
This goes into _post_process just like normal uploads, that's not a problem?

Reply:
It won't be a problem. `container_internal_file` is used to build WS object. We won't build Assembly object. The assembly object associate with genome is build from the genbank file.

Reply:
I'm not sure I understand - container_internal_file is the location for the file inside the docker container. That's not used to build the ws object

Reply:
:+1: